{{cookiecutter.project_name}}/.github/workflows/test.yaml

```diff
@@ -18,44 +18,61 @@ defaults:
     shell: bash -euo pipefail {0}
 
 jobs:
+  get-environments:
+    runs-on: ubuntu-latest
+    outputs:
+      matrix: ${{ steps.get-envs.outputs.matrix }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          filter: blob:none
+          fetch-depth: 0
+      - name: Install uv
+        uses: astral-sh/setup-uv@v5
+      - name: Get test environments
+        id: get-envs
+        run: |
+          MATRIX_JSON=$(uvx hatch env show --json | jq -c 'to_entries
+            | map(
+                select(.key | startswith("hatch-test"))
+                | {
+                    env_name: .key,
+                    name: (if (.key | contains("pre")) then .key + " (PRE-RELEASE DEPENDENCIES)" else .key end),
+                    python: .value.python
+                  }
+              )')
+          echo "matrix=${MATRIX_JSON}" | tee $GITHUB_OUTPUT
+
   test:
+    needs: get-environments
     runs-on: ${{ matrix.os }}
 
     strategy:
       fail-fast: false
       matrix:
-        include:
-          - os: ubuntu-latest
-            python: "3.10"
-          - os: ubuntu-latest
-            python: "3.12"
-          - os: ubuntu-latest
-            python: "3.12"
-            pip-flags: "--pre"
-            name: PRE-RELEASE DEPENDENCIES
-
-    name: ${{ matrix.name }} Python ${{ matrix.python }}
-
-    env:
-      OS: ${{ matrix.os }}
-      PYTHON: ${{ matrix.python }}
+        os: [ubuntu-latest]
+        include: ${{ fromJson(needs.get-environments.outputs.matrix) }}
```

Comment:
I don't understand why this doesn't work. The test job in `cookiecutter-scverse-instance` only runs for a single environment even though the json looks exactly as I would expect: 

![image](https://github.com/user-attachments/assets/448000ea-2831-4575-8fab-3a76bc323879)

https://github.com/scverse/cookiecutter-scverse-instance/pull/195

Reply:
Only thing I can see is that you typed `fromJson` instead of `fromJSON`

Reply:
Ah, the issue is the overcomplicated way `include` works: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/running-variations-of-jobs-in-a-workflow#expanding-or-adding-matrix-configurations

Reply:
I think in order to do what we want, we need to define our matrix using both `os` and `env_name` arrays, and then “[expand](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/running-variations-of-jobs-in-a-workflow#example-expanding-configurations)” the configurations based on `env_name`.

```yaml
matrix:
  os: [ubuntu-latest]
  env_name: [...]
  include:
  - env_name: ...
    name: ...
    python: ...
  ...
```